src/pybind/mgr/dashboard/tools.py

```diff
@@ -521,11 +534,217 @@ def run(self):
+            'metadata': t.metadata,
+            'begin_time': "{}Z".format(datetime.fromtimestamp(t.begin_time).isoformat()),
+            'end_time': "{}Z".format(datetime.fromtimestamp(t.end_time).isoformat()),
+            'duration': t.duration,
+            'progress': t.progress,
+            'success': not t.exception,
+            'ret_value': t.ret_value,
+            'exception': t.exception
+        } for t in fn_t]
+
+
+class AsyncTask(object):
+    # pylint: disable=too-many-instance-attributes
+
+    class ExecutorThread(threading.Thread):
+        def __init__(self, task):
+            super(AsyncTask.ExecutorThread, self).__init__()
+            self._task = task
+            self.event = threading.Event()
+
+        # pylint: disable=broad-except
+        def run(self):
+            TaskManager._task_local_data.task = self._task
+            try:
+                self._task.set_progress(0)
+                self._task.begin_time = time.time()
+                val = self._task.fn(*self._task.fn_args, **self._task.fn_kwargs)
+                self._task.end_time = time.time()
+            except Exception as ex:
+                logger.exception("Error while calling %s", self._task)
+                with self._task.lock:
+                    self._task.end_time = time.time()
+                    self._task.ret_value = None
+                    self._task.exception = ex
+            else:
+                with self._task.lock:
+                    self._task.duration = self._task.end_time - self._task.begin_time
+                    self._task.ret_value = val
+                    self._task.exception = None
+                    self._task.set_progress(100, True)
+                logger.debug("execution of %s finished in: %s s", self._task, self._task.duration)
+            finally:
+                NotificationQueue.new_notification('cd_task_finished', self._task)
+                self.event.set()
+
+    def __init__(self, namespace, metadata, fn, args, kwargs):
+        self.namespace = namespace
+        self.metadata = metadata
+        self.fn = fn
+        self.fn_args = args
+        self.fn_kwargs = kwargs
+        self.executor_thread = None
+        self.event = threading.Event()
+        self.progress = None
+        self.ret_value = None
+        self.begin_time = None
+        self.end_time = None
+        self.duration = 0
+        self.exception = None
+        self.lock = threading.Lock()
+
+    def __hash__(self):
+        return hash((self.namespace, tuple(sorted(self.metadata.items()))))
+
+    def __eq__(self, other):
+        return self.namespace == self.namespace and self.metadata == self.metadata
+
+    def __str__(self):
+        return "Task(ns={}, md={})" \
+               .format(self.namespace, self.metadata)
+
+    def _run(self):
+        with self.lock:
+            if self.executor_thread is None:
+                self.executor_thread = AsyncTask.ExecutorThread(self)
```

Comment:
Threads are needed for tasks that use blocking apis, but I'm not sure about using them for tasks in general.

For tasks that use mon commands, it's nice to avoid the thread and just call back into them when the command response messages come in -- that way it's a lot easier to implement cancelling or otherwise interrupting the task.  That's handy for putting a cancel button in the UI, but I think we're also going to want it for cases where we might want to cancel some tasks in response to health issues.

Reply:
When I first started thinking on the solution for the asynchronous tasks, I thought about this problem, but then when I was developing the solution I completely forgot to add the support for calling functions that were already asynchronous. Thanks for raising the issue.

My idea is to keep the `TaskManager` interface, so from the API consumers perspective they will only see "tasks". I think I already know how to implement the support for async API calls without changing too much the current interface. I'll push the implementation to this PR and then notify you for the review.